cadence/values.yaml

```diff
@@ -57,6 +57,65 @@ server:
   affinity: {}
 
   config:
+    # clusterMetadata contains the settings of the Cadence clusters
+    # participating in a Cadence cluster group.
+    clusterMetadata:
+      # enableGlobalDomains when set to true allows creating global domains
+      # which are replicated between all clusters and whose tasks are served by
+      # the active cluster in the replication.
+      #
+      # WARNING: This value **MUST NOT** be changed after launching the Cadence
+      # cluster group.
+      enableGlobalDomain: true
+
+      # failoverVersionIncrement is the increment of each cluster version when
+      # failover happens.
+      #
+      # WARNING: For each cluster, (initialFailoverVersion +
+      # i*failoverVersionIncrement) **MUST** result in a unique value.
+      failoverVersionIncrement: 10
+
+      # masterClusterName is the name of the only cluster which can register
+      # and update a domain, although each cluster can do domain failover.
+      #
+      # WARNING: This value **MUST NOT** be changed after launching the Cadence
+      # cluster group.
+      masterClusterName: "cluster-0"
+
+      # currentClusterName is the name of the current active cluster for the
+      # global domain.
+      currentClusterName: "cluster-0"
+      
+      # clusterInformation contains the information about each cluster which
+      # participates in a Cadence cluster group.
+      #
+      # WARNING: cluster names and initialFailoverVersion values **MUST** be
+      # unique.
+      #
+      # WARNING: For each cluster, (initialFailoverVersion +
+      # i*failoverVersionIncrement) **MUST** result in a unique value.
+      clusterInformation:
+        cluster-0:
+          # enabled defines whether the cluster is enabled.
+          enabled: true
+          # initialFailoverVersion is the failover version a cluster starts at
+          # when launched.
+          #
+          # WARNING: This value **MUST** be unique and (initialFailoverVersion
+          # + i*failoverVersionIncrement) **MUST** result in a unique value.
+          initialFailoverVersion: 0
+          # rpcName defines the name of the Cadence frontend remote service to
+          # use in a Cadence cluster.
+          rpcName: cadence-frontend
+          # rpcAddress defines the address (host:port, host can be a DNS name)
+          # of the Cadence frontend remote service to use in a Cadence cluster.
+          rpcAddress: cadence:7933
```

Comment:
Would `cadence:7933` work by default? Shouldn't it be something like "service/<releaseName>-frontend"?

Reply:
In a single cluster setup, this value should be the same as publicClient.hostPort: 
https://github.com/banzaicloud/banzai-charts/blob/d8d883e026bf1b492b1aeb89383eab8105e5ab2b/cadence/templates/server-configmap.yaml#L185

Some background:
Ideally we should remove publicClient as the same information is provided by this rpcAddress. 
see issue: https://github.com/uber/cadence/issues/3984

Reply:
I made a mistake here, this was supposed to be `cadence-frontend`, [corrected it](https://github.com/banzaicloud/banzai-charts/compare/ecca195af76a2ed2d5e4ef3bf31d6c7567fe35ce..e18536f38ea7d0e1491f409fc4359e11c22178d8).
(I don't yet understand how this passed the manual acceptance tests, so I'm redoing it.)

So the named address works with the current chart configuration, because
1. we define a [K8s service](https://kubernetes.io/docs/concepts/services-networking/service/), named [`cadence-frontend`](https://github.com/banzaicloud/banzai-charts/blob/feat/parametrize-cadence-cluster-names/cadence/templates/server-service.yaml#L4),

2. which [serves the corresponding `cadence-frontend` pod](https://github.com/banzaicloud/banzai-charts/blob/feat/parametrize-cadence-cluster-names/cadence/templates/server-service.yaml#L63), and

3. the service is assigned a [`ClusterIP`](https://kubernetes.io/docs/concepts/services-networking/service/#publishing-services-service-types) (internal to the K8s cluster) which can proxy the service endpoint's requests to the currently running served pods (through endpoint objects, but that's detail),
```shell
-> kubectl --kubeconfig "${BANZAI_INSTALLER_WORKSPACE_KUBECONFIG}" --namespace banzaicloud get service cadence-frontend

NAME               TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)    AGE
cadence-frontend   ClusterIP   10.104.122.83   <none>        7933/TCP   63m
```

4. [K8s automatically assigns DNS records to services and pods](https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/) and resolves the `cadence-frontend` host to `cadence-frontend.cadence-namespace.svc.cluster.local` for pods coming from the same namespace (for pods in other namespaces, `cadence-frontend.cadence-namespace` namespace-qualified host is required to be used) and the full DNS resolves to the `ClusterIP` assigned to the service,
```shell
-> kubectl --kubeconfig "${BANZAI_INSTALLER_WORKSPACE_KUBECONFIG}" --namespace banzaicloud run curl --image=radial/busyboxplus:curl -i --tty

[ root@curl:/ ]$ nslookup cadence-frontend
Server:    10.96.0.10
Address 1: 10.96.0.10 kube-dns.kube-system.svc.cluster.local

Name:      cadence-frontend
Address 1: 10.104.122.83 cadence-frontend.banzaicloud.svc.cluster.local
```

5. which other services can use afterwards.
```shell
-> kubectl --kubeconfig "${BANZAI_INSTALLER_WORKSPACE_KUBECONFIG}" --namespace banzaicloud logs cadence-frontend-579d44c8bc-mlchf | grep cadence-frontend

# Begin: 2021-08-02T22:56:56.566525000+02:00

{"level":"info","ts":"2021-08-02T20:53:22.184Z","msg":"Updated dynamic config","service":"cadence-frontend","logging-call-at":"file_based_client.go:258"}
{"level":"info","ts":"2021-08-02T20:53:22.185Z","msg":"Listening for TChannel requests","service":"cadence-frontend","address":"10.244.0.39:7933","logging-call-at":"rpc.go:90"}
{"level":"info","ts":"2021-08-02T20:53:22.185Z","msg":"Listening for GRPC requests","service":"cadence-frontend","address":"10.244.0.39:7833","logging-call-at":"rpc.go:101"}
{"level":"info","ts":"2021-08-02T20:53:22.187Z","msg":"First loading dynamic config","service":"cadence-frontend","key":"system.advancedVisibilityWritingMode,clusterName:cluster-0","value":"off","default-value":"off","logging-call-at":"config.go:93"}
{"level":"info","ts":"2021-08-02T20:53:22.187Z","msg":"Creating RPC dispatcher outbound","service":"cadence-frontend","address":"cadence-frontend:7933","logging-call-at":"clientBean.go:277"}
{"level":"info","ts":"2021-08-02T20:53:22.188Z","msg":"Starting service frontend","service":"cadence-frontend","logging-call-at":"server.go:219"}
{"level":"info","ts":"2021-08-02T20:53:22.189Z","msg":"Creating RPC dispatcher outbound","service":"cadence-frontend","address":"cadence-frontend:7933","logging-call-at":"clientBean.go:277"}
{"level":"info","ts":"2021-08-02T20:53:22.191Z","msg":"Add new peers by DNS lookup","service":"cadence-frontend","address":"cadence-frontend","addresses":"[10.104.122.83:7933]","logging-call-at":"clientBean.go:321"}
{"level":"info","ts":"2021-08-02T20:53:22.191Z","msg":"Add new peers by DNS lookup","service":"cadence-frontend","address":"cadence-frontend","addresses":"[10.104.122.83:7933]","logging-call-at":"clientBean.go:321"}

-> kubectl --kubeconfig "${BANZAI_INSTALLER_WORKSPACE_KUBECONFIG}" --namespace banzaicloud logs cadence-worker-6754f84875-gx6b2 | grep cadence-frontend

# Begin: 2021-08-02T22:54:46.643479000+02:00

{"level":"info","ts":"2021-08-02T19:55:22.937Z","msg":"Creating RPC dispatcher outbound","service":"cadence-worker","address":"cadence-frontend:7933","logging-call-at":"clientBean.go:277"}
{"level":"info","ts":"2021-08-02T19:55:22.938Z","msg":"Creating RPC dispatcher outbound","service":"cadence-worker","address":"cadence-frontend:7933","logging-call-at":"clientBean.go:277"}
{"level":"info","ts":"2021-08-02T19:55:22.941Z","msg":"Add new peers by DNS lookup","service":"cadence-worker","address":"cadence-frontend","addresses":"[10.104.122.83:7933]","logging-call-at":"clientBean.go:321"}
{"level":"info","ts":"2021-08-02T19:55:22.941Z","msg":"Add new peers by DNS lookup","service":"cadence-worker","address":"cadence-frontend","addresses":"[10.104.122.83:7933]","logging-call-at":"clientBean.go:321"}
```

Reply:
Gonna process your second comment in a minute.
My comment above was only meant to reply to your first comment.

Reply:
Yes, your second comment about `publicClient` also makes sense.

I haven't revised that part yet in light of the configuration changes, so I'm going to do it now.
The defaulting there is completely unnecessary, because both chart values are always set, so currently the real template value of it is
```yaml
publicClient:
      hostPort: "{{ include "cadence.componentname" (list . "frontend") }}:{{ .Values.server.frontend.service.port }}"
```
which equals `cadence-frontend:7933` by default.

I'm currently thinking about the feasibility of the multi-cluster setup and how the DNS could/should be specified in that case.

Reply:
Sounds good. 

More background.  
It's a little bit hard to confirm this rpcAddress. In a single cluster setup, the only place that address is being used is in a batcher system workflow:
https://github.com/uber/cadence/blob/9a46d9d483cfdb8fd5ca77d1f448d79d78b0e791/service/worker/batcher/workflow.go#L240
Which only get triggered if you start a batch job by CLI:
```
cadence --do samples-domain wf batch start --query "WorkflowType='main.SampleParentWorkflow' AND CloseTime=missing" --reason "test" --bt signal --sig testname
This batch job will be operating on 5 workflows.
Please confirm[Yes/No]:yes
{
    "jobID": "<batch-job-id>",
    "msg": "batch job is started"
}
```
See https://cadenceworkflow.io/docs/cli/#signal-cancel-terminate-workflows-as-a-batch-job

The worker service is started correctly because it always use the `publicClient.hostPort` config to connect to frontend:
e.g.:
https://github.com/uber/cadence/blob/master/service/worker/service.go#L257

Which means users won't know this rpcAddress is wrong until they use the batcher system workflow feature. 


> which equals cadence-frontend:7933 by default.

Yeah I think by default it's correct. But when I used this command to install my cluster: ` helm install cadence-qa-1  -f ...`
I got 
```
    publicClient:
      hostPort: "cadence-qa-1-frontend:7933"
```
So I guess `"cadence.componentname"` is using the releaseName for the prefix.

Reply:
Yes it does.
By default I both meant the default release name and the default server config frontend port.

Reply:
I would say currently the chart doesn't support multi-cluster Cadence setups.

It requires a refactor on the service definition to support ingresses, loadbalancers, external names, etc.
And the discussed correction on the `publicClient.hostPort` and for it to be synchronized to the K8s cluster's correspnding Cadence cluster.

To me it seems like it would be best added in a different PR, because it sounds like a bigger change to me on its own already.

Reply:
A separate Pr for multi cluster support sounds good to me. 

To confirm, we are going to do the same as publicClient right? Instead of hardcoded to cadence-fronted:7933 which will not work for customed release name

Reply:
Please note that the values.yaml isn't supposed to be a hard-coded solution for all use-cases, it provides a baseline which users should customize to their individual use-cases, including custom release names.
So for the time being I believe it is a fair expectation of the user to update the `values.yaml.server.config.clusterMetadata.clusterInformation[*].rpcAddress` values, even if not really convenient.

I don't see a perfect solution yet how to make it possible to configure both custom DNSes in the values.yaml and also use the release name, either the user has to configure it manually to match the release name or we lose the possibility to customize it.

At this point, I'm even considering reverting part of this PR and not allow rpcAddress customization for the time being which would solve the issue of different release names.

I have to take a timeout and think about this for a bit what should be allowed to be customized and how.

Reply:
Ahh I got it now. Just realized that this is values.yaml not template so we can’t do the same as publicClient. What you said makes sense to me. I think at least having comments about it on changing the value should be fine for me. But if you can come up with something else that’s also great. 

Sorry for my back and forth comments on this. Thanks for the great work  and being patient with me😬

Reply:
Sure, the comment is a good idea, gonna do it later today (I'm bit swamped at work).

Also I'm thinking about a possible solution to sync the `publicClient` with one of the `clusterInformation.*.rpcAddress`. Gonna update the PR once I have a solution ready.

It's okay, I should be apologizing for causing the bugs in the first place and also I should be grateful to you, because I'm only scratching the surface of Cadence at this point, but your guidance helped a lot.

Reply:
Resolving, see https://github.com/banzaicloud/banzai-charts/pull/1278#issuecomment-899801805 for an update on this.